test/libsolidity/semanticTests/abiEncoderV2/dirty_memory_dynamic_array_and_calldata_string.sol

```diff
@@ -0,0 +1,72 @@
+pragma abicoder v2;
+
+contract C
+{
+    struct Data
+    {
+        uint256[] a;
+        string b;
+    }
+
+    function makeDirty() internal
+    {
+        uint memStart = 0;
+        assembly { memStart := mload(0x40) }
+        uint memEnd = memStart + 0x800;
+
+        for (uint pos = memStart; pos < memEnd; pos += 0x20)
+            assembly { mstore(pos, not(0)) }
+    }
+
+    function makeTestData() internal returns (Data memory data)
+    {
+        data.a = new uint256[](1);
+        data.a[0] = 0xFF;
+        data.b = "123456";
+    }
+
+    function makeTestArray() internal returns (uint256[] memory array)
+    {
+        array = new uint256[](2);
+        array[0] = 0xAF;
+        array[1] = 0xBF;
+    }
+
+    function testDirtyCalldata() public returns (bytes memory)
+    {
+        makeDirty();
+
+        return this.calldataTest(makeTestData(), makeTestArray());
+    }
+
+    function testDirtyMemory() public returns (bytes memory)
+    {
+        makeDirty();
+
+        return this.memoryTest(makeTestData(), makeTestArray());
+    }
+
+    function calldataTest(Data calldata data, uint256[] calldata a) public pure returns (bytes memory)
+    {
+        return abi.encode(data, a);
+    }
+
+    function memoryTest(Data memory data, uint256[] calldata a) public pure returns (bytes memory)
+    {
+        return abi.encode(data, a);
+    }
```

Comment:
Bad news. This test no longer detects the missing cleanup :(

Just tested it locally by replacing `bytesOrString` in `ABIFunctions.cpp` with `false` and the test still passes. Isn't this because these functions are external so the dirtying does not affect what happens inside them?

Reply:
With ``bytesOrString == false`` the new version *still* won't perform the old cleanup - that way you call it as if it was a ``uint[]`` or ``bytes32[]`` array, for which we generally removed the cleanup. That's the reason, isn't it? Or did I get it wrong what you mean :-)?

Reply:
Yeah, I think I misunderstood, let me have a closer look...

Reply:
> With `bytesOrString == false` the new version still won't perform the old cleanup

That's the point. It's a separate thing from the bug. One of the early versions of this PR was basically equivalent to `bytesOrString == false`, which removes the cleanup in the `bytes` and `string` case. This means that byte arrays were not being cleaned. I noticed that and asked for a test covering that - since nothing failed which means we were missing coverage.

Reply:
For reference, here's my original comment where I explained that: https://github.com/ethereum/solidity/pull/13282#discussion_r926963789

Reply:
The test gives me a headache :-). If it's just the cleanup when abi-encoding a string into previously dirty memory, this should be easier IMO:
```
ontract C {
  function f(string calldata x) external returns (bytes memory r) {
        uint mptr;
        assembly {
                // dirty memory
                mptr := mload(0x40)
                for { let i := mptr } lt(i, add(mptr, 0x0100)) { i := add(i, 32) }
                {
                    mstore(i, sub(0,1))
                }
        }
        r = abi.encode(x);
        assembly {
                // artificially increase length of r to multiple of 32
                mstore(r, and(add(mload(r), 31), not(31)))
                // assert that we dirtied the memory that was encoded to
                if iszero(eq(mptr, r)) { revert(0,0) }
        }
  }
  function test() external returns (bytes memory) {
        return this.f("abc");
  }
}
// ----
// test() -> 0x20, 0x60, 0x20, 3, "abc"
```

Reply:
I pushed a commit that adds @ekpyron's test and verified that it's effective. Not sure what to do about the existing one. If it's still broken, probably better to remove it.